File: sem01/java/maga/src/main/java/org/khomenko/maga/stream/Iris.java
package org.khomenko.maga.stream;

import org.khomenko.maga.csv.CsvColumn;

public class Iris {
    @CsvColumn
    private Double sepalLength; // длина чашелистника

    @CsvColumn
    private double sepalWidth; // ширина чашелистника

    @CsvColumn
    private double petalLength; // длина лепестка

    @CsvColumn
    private double petalWidth; // ширина лепестка

    @CsvColumn
    private String species; // вид

    public enum Petal {
        SMALL, MEDIUM, LARGE,
    }

    public static Petal classifyByPatel(Iris iris) {
        double patelSquare = iris.getPetalWidth() * iris.getPetalLength();
        if (patelSquare < 2.0) {
            return Petal.SMALL;
        } else if (patelSquare < 5.0) {
            return Petal.MEDIUM;
        }
        return Petal.LARGE;
    }

    public Iris() {}

    public Iris(double sepalLength, double sepalWidth, double petalLength, double petalWidth, String species) {
        this.sepalLength = sepalLength;
        this.sepalWidth = sepalWidth;
        this.petalLength = petalLength;
        this.petalWidth = petalWidth;
        this.species = species;
    }

    public double getSepalLength() {
        return sepalLength;
    }

    public double getSepalWidth() {
        return sepalWidth;
    }

    public double getPetalLength() {
        return petalLength;
    }

    public double getPetalWidth() {
        return petalWidth;
    }

    public String getSpecies() {
        return species;
    }

    @Override
    public boolean equals(Object other) {
        if (other == null) {
            return false;
        }

        if (this.getClass() != other.getClass()) {
            return false;
        }

        var iris = (Iris)other;

        return getPetalLength() == iris.getPetalLength() &&
                getPetalWidth() == iris.getPetalWidth() &&
                getSepalLength() == iris.getSepalLength() &&
                getSepalWidth() == iris.getSepalWidth() &&
                getSpecies().equals(iris.getSpecies());
    }

    @Override
    public String toString() {
        return "Iris{" +
                "sepalLength=" + sepalLength +
                ", sepalWidth=" + sepalWidth +
                ", petalLength=" + petalLength +
                ", petalWidth=" + petalWidth +
                ", species='" + species + "'" +
                '}';
    }
}

File: sem01/scala/maga/src/main/scala/org/khomenko/maga/collections/Iris.scala
package org.khomenko.maga.collections

import scala.util.Try

object PetalSize extends Enumeration {
  type PetalSize = Value
  val Small, Medium, Large = Value
}

case class Iris(sepalLength: Double, sepalWidth: Double,
                petalLength: Double, petalWidth: Double,
                species: String) {
  def getPetalSquare(): Double = petalLength * petalWidth

  def getPetalSize(): PetalSize.PetalSize = {
    val petalSquare = petalLength * petalWidth
    if (petalSquare < 2.0) {
      PetalSize.Small
    } else if (petalSquare < 5.0) {
      PetalSize.Medium
    } else {
      PetalSize.Large
    }
  }
}

object IrisUtils {
  implicit class StringToIris(str: String) {
    def toIris: Option[Iris] = str.split(",") match {
      case Array(a, b, c, d, e) if isDouble(a) && isDouble(b) && isDouble(c) && isDouble(d) =>
        Some(
          Iris(
            a.toDouble,
            b.toDouble,
            c.toDouble,
            d.toDouble,
            e))
      case others => None
    }

    def isDouble(str: String): Boolean = Try(str.toDouble).isSuccess
  }
}